Misc

jschueller opened the pull request:
- Metropolis final touch
- new Collection::select(Indices) method
- https://github.com/openturns/openturns/issues/1946

jschueller commented:
cc @josephmure @MerlinKeller for the metropolis stuff

josephmure commented on python/src/RandomVectorMetropolisHastings_doc.i.in:
```diff
@@ -31,9 +31,9 @@ Examples
 >>> initialState = [3.0]
->>> sampler = ot.RandomVectorMetropolisHastings(source, initialState)
+>>> sampler = ot.RandomVectorMetropolisHastings(randomVector, initialState)
 >>> sampler.setBurnIn(20)
 >>> sampler.setThinning(2)
```
```suggestion
```
Makes no sense in this context: we are directly sampling from the random vector! No burn-in or thinning is required.

josephmure commented on python/src/RandomVectorMetropolisHastings_doc.i.in:
```diff
@@ -31,9 +31,9 @@ Examples
+>>> randomVector = ot.RandomVector(ot.Normal())
 >>> initialState = [3.0]
->>> sampler = ot.RandomVectorMetropolisHastings(source, initialState)
+>>> sampler = ot.RandomVectorMetropolisHastings(randomVector, initialState)
```
I am not sure we need an example that trivial: `sampler.getRealization()` is here equivalent to `randomVector.getRealization()`!

josephmure commented on python/src/RandomVectorMetropolisHastings_doc.i.in:
```diff
@@ -42,22 +42,22 @@ We can also link the parameters of the random vector to the state of the chain,
 here the parameters of the random vector being the ones of the underlying distribution:
 
 >>> slf = ot.SymbolicFunction(['x'], ['0.0', '0.1'])
```
```suggestion
>>> linkFunction = ot.SymbolicFunction(['x'], ['x', '0.1'])
```
That way `sampler` becomes a random walk, which is a little more interesting than having the parameters be independent from the state of the Markov chain.

josephmure commented on python/src/RandomVectorMetropolisHastings_doc.i.in:
```diff
@@ -42,22 +42,22 @@ We can also link the parameters of the random vector to the state of the chain,
 
 >>> slf = ot.SymbolicFunction(['x'], ['0.0', '0.1'])
->>> sampler = ot.RandomVectorMetropolisHastings(source, initialState, [0], slf)"
+>>> sampler = ot.RandomVectorMetropolisHastings(randomVector, initialState, [0], slf)"
```
```suggestion

The link function makes the first parameter of the normal distribution (the mean)
equal to the current value of the Markov chain.
Its standard deviation remains constant: :math:`0.1`.
That way we construct a random walk with normal steps of standard deviation :math:`0.1`.

>>> sampler = ot.RandomVectorMetropolisHastings(randomVector, initialState, [0], linkFunction)"
```

josephmure commented on python/src/RandomVectorMetropolisHastings_doc.i.in:
```diff
@@ -42,22 +42,22 @@ We can also link the parameters of the random vector to the state of the chain,
 >>> slf = ot.SymbolicFunction(['x'], ['0.0', '0.1'])
->>> sampler = ot.RandomVectorMetropolisHastings(source, initialState, [0], slf)"
+>>> sampler = ot.RandomVectorMetropolisHastings(randomVector, initialState, [0], slf)"
 
```
I think the RandomVectorMetropolisHastings API doc page should mention that the class is designed to work with the `Gibbs` class and link to the "Gibbs sampling of the posterior distribution" example.

jschueller replied:
It's simpler to use "See Also" for the relevant classes

josephmure reviewed (commented):
Great! I particularly like the new Collection::select method.

My review is not finished. I will get back to it later.

josephmure commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,583 @@
+X = np.array(ot.ComposedDistribution([ot.Dirac(1.0), 10.0 * ot.Normal()]).getSample(n))
+print(X[:,1])
+
+# add random exponentialy distributed precisions
+Q = np.ones((n,1)) + np.array( ot.Exponential().getSample(n) ).reshape(-1,1)
```
```suggestion
# %%
# Make the precision matrix :math:`Q` a diagonal matrix and sample
# its diagonal coefficients from an :class:`~openturns.Exponential` distribution.

Q = np.ones((n,1)) + ot.Exponential().getSample(n)
```

2-d numpy array + Sample = 2-d numpy array so the code can be simplified.

josephmure commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,583 @@
+p = len(theta_true)
+tau_true = .1
+
+# Y conditional distribution
```
```suggestion
# %%
# First sample uncensored, and then censored observation data.

```

josephmure commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,583 @@
+Y_sim = mean_true + [x[0] for x in ot.Normal().getSample(n)] * std_true
+Yobs_sim = np.round(Y_sim/delta)*delta
+
+
+# %%
```
```suggestion
# %%
# Plot the simulated dataset.

```

josephmure commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,583 @@
+#    \theta | \tau \sim \mathcal N_{p}(\mu_0; \tau^{-1}\Sigma_0) {\bf 1}_{\{\theta_{\min} \leq \theta \leq \theta_{\max}\}} ;\\
+#    \tau \sim \mathcal G(a_0, b_0){\bf 1}_{\{\tau_{\min} \leq \tau \leq \tau_{\max}\}},
+# 
+# where inequality are taken componentwise. When all hyperparameters go to
```
```suggestion
# where inequalities are taken componentwise. When all hyperparameters go to
```

josephmure reviewed (commented):
Second part of my review. The last part will come as soon as possible.

jschueller commented on .circleci/run_docker_linux.sh:
```diff
@@ -10,7 +10,7 @@ env
 if test "${CIRCLE_BRANCH}" = "master"
 then
-  export SOURCE_DATE_EPOCH=1640991600
+  export SOURCE_DATE_EPOCH=1640995200
```
we want to have 2022 in the sphinx generation date

jschueller commented on lib/src/Uncertainty/Bayesian/RandomVectorMetropolisHastings.cxx:
```diff
@@ -76,55 +73,45 @@ RandomVectorMetropolisHastings* RandomVectorMetropolisHastings::clone() const
-  adv.loadAttribute("source_", source_);
-  adv.loadAttribute("sourceLinkFunction_", sourceLinkFunction_);
+  adv.loadAttribute("randomVector_", randomVector_);
+  adv.loadAttribute("linkFunction_", linkFunction_);
```
absolutely

josephmure commented on python/src/RandomVectorMetropolisHastings_doc.i.in:
```diff
@@ -31,33 +38,37 @@ Examples
 
-%feature("docstring") OT::RandomVectorMetropolisHastings::setSource
+%feature("docstring") OT::RandomVectorMetropolisHastings::setRandomVector
 "Set the proposal distribution.
```
```suggestion
"Set the random vector.
```

josephmure replied:
Note changes also need to be made to the parameter description immediately below this line, but GitHub wouldn't let me reach those lines.

josephmure commented on python/src/RandomVectorMetropolisHastings_doc.i.in:
```diff
@@ -5,63 +5,82 @@ Refer to :ref:`bayesian_calibration`, :ref:`metropolis_hastings`.
 Examples
 --------
 >>> import openturns as ot
 >>> import math as m
 >>> ot.RandomGenerator.SetSeed(0)
```
```suggestion
```
Do we really need those lines?

regislebrun replied:
I don't know for the first one, but for the second one it is mandatory if you want to have a robust output, otherwise if another class using RandomGenerator is inserted in the documentation before RandomVectorMetropolisHastings, then the state of RandomGenerator will be different when the snippets of code will be executed, producing a different output than the reference one.

josephmure replied:
No output is printed, so I don't think it matters.

jschueller replied:
the math import is useless, but lets avoid surprises with the state of the rng

josephmure replied:
Works for me.

josephmure commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,591 @@
+
+# %%
+# Draw pairplots of the posterior sample.
+# sphinx_gallery_thumbnail_number = 3
```
Here, "sphinx_gallery_thumbnail_number = 3" appears as plain text rather than code. Note that the GitHub suggestion where I proposed introducing this also introduced a blank line above
```python
# sphinx_gallery_thumbnail_number = 3
```
precisely to avoid this.

jschueller replied:
ok, ill check the output

josephmure reviewed (changes requested):
The "plot_gibbs_simus.py" example looks good overall, but there are quite a few formatting errors. Most of them concern section or subsection titles. For example, the "Remarks" subsection currently continues way beyond the bullet list of remarks and contains the Python code that produces the data.

It would take longer to list those errors than to correct them so please look carefully at the output HTML file. If you don't want to compile all the doc, you can use the "docfast.py" utility ;)

Also, I think you have forgotten some of the suggestions from the second part of my review: they have been hidden by GitHub and you need to click on "Load more" to see them.

regislebrun commented:
It is only a good practice for classes involving random sampling. If you add some output afterward it will save you some fead scratching. But feel free to remove it if you prefer.

josephmure commented:
@MerlinKeller Can you have a look at Metropolis-related files, especially "plot_gibbs_simus.py"? Unfortunately, the "build_linux_artifact Details" link no longer works: it seems that CircleCI changed the structure of the URL of its artifacts (that is, the HTML doc files in our case). However, you can still access them by clicking on the "build_linux Details" button, and then on the "Artifacts" tab in the CircleCI screen.

jschueller commented:
> Remarks



> The "plot_gibbs_simus.py" example looks good overall, but there are quite a few formatting errors. Most of them concern section or subsection titles. For example, the "Remarks" subsection currently continues way beyond the bullet list of remarks and contains the Python code that produces the data.
> 
> It would take longer to list those errors than to correct them so please look carefully at the output HTML file. If you don't want to compile all the doc, you can use the "docfast.py" utility ;)
> 
> Also, I think you have forgotten some of the suggestions from the second part of my review: they have been hidden by GitHub and you need to click on "Load more" to see them.

ok, I fixed the sections, added a new one for the dataset generation

josephmure commented:
> The `select` method is a great simplification.

I agree, and I will endeavor to update the IndependentMetropolisHastings PR (#1976) as soon as this is merged into master.

mbaudin47 commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,598 @@
+# .. math::
+# 
+#    \pi(Y,\theta,\tau|Y^{obs})
+#    \propto
```
Please make the `propto` symbols aligned, to make the equations more readable:
![image](https://user-images.githubusercontent.com/31351465/160208140-a2723ff2-3b4d-4a61-9847-7e335d134f41.png)

mbaudin47 reviewed (changes requested):
The linear regression with interval censored observations, but does not seem to be ready for integration in my point of view.

josephmure commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,598 @@
+
+class BoxConstrainedNormal(ot.PythonDistribution):
+    """
+    Multivariate normal distribution
```
No, @MerlinKeller tried it and it did not work as expected.

MerlinKeller replied:
More to the point, the box-constrained multivariate normal is NOT equivalent to a join law with normal copula and truncated normal marginals.

josephmure commented:
Given the numerous changes requested in `plot_gibbs_simus.py`, I suggest removing the file from this PR. An other PR, dedicated to this example, can then be created. Such a PR could target the 1.19 branch even during the release candidate phase, couldn't it? It would not modify the actual library.

Moreover, the current PR could then more easily be merged, which would be welcome since #1976 depends on it.

jschueller commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,598 @@
+# and :math:`\tau` an unknown precision parameter
+# quantifying the variability of the observed phenomenon.
+#
+# 1.1. Likelihood of the IPE model
```
no clue

jschueller commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,598 @@
+"""
+
+# %%
+# 1. Model formulation
```
I guess not

jschueller commented on python/doc/examples/calibration/bayesian_calibration/plot_gibbs_simus.py:
```diff
@@ -0,0 +1,598 @@
+import openturns as ot
+from openturns.viewer import View
+ot.Log.Show(ot.Log.NONE)
+ot.RandomGenerator.SetSeed(11123)
```
actually no, not in examples as its not a new python process

jschueller commented:
> Given the numerous changes requested in `plot_gibbs_simus.py`, I suggest removing the file from this PR. An other PR, dedicated to this example, can then be created. Such a PR could target the 1.19 branch even during the release candidate phase, couldn't it? It would not modify the actual library.
> 
> Moreover, the current PR could then more easily be merged, which would be welcome since #1976 depends on it.
> 
> EDIT: @jschueller you might however want to apply the GitHub suggestions to `plot_gibbs_simus.py` immediately so they are not lost.

ok, I dropped it

josephmure reviewed (approved):
When this is merged, PR #1976 will need to rebase, because it depends on some of the changes made here.

jschueller commented:
@regislebrun or @sofianehaddad I'd like to merge this one too, since #1976 depends on it
